radix-engine/src/fee/fee_reserve.rs

```diff
@@ -209,6 +239,25 @@ impl SystemLoanFeeReserve {
 }
 
 impl FeeReserve for SystemLoanFeeReserve {
+    fn consume_deferred(
```

Comment:
Why do need a dedicated method for this wit duplicated code?

Reply:
Because:
* At one point, I wanted to give them different semantics -- ie create a precondition that you weren't allowed to call `consume_deferred` after any calls to `consume` -- before realising it wasn't quite necessary. But I preferred the clarity of the API of the separate methods so kept it.
* The true/false flag was a bit confusing  - it seemed to suggest that it was a choice. But `consume_deferred` is not called from inside the engine after original set-up, and really, shouldn't be. 
* `consume` is on a hot code path so reducing the logic in the method is reasonable
* Boolean parameters also are best avoided as a very general rule: https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3 - and `consume` and `consume_deferred` didn't actually share much code

Reply:
Is it just because of missing documentation? 

I'm more worried that we might forget to update one place when updating another, especially for anyone who's new to codebase. 

Not convinced we would save much from a single branching instruction.

Reply:
I think the methods are fundamentally different from an abstraction perspective - compare `consume_deferred`:
```
        if amount == 0 {
            return Ok(());
        }

        let units_consumed = checked_multiply(amount, multiplier)?;

        checked_assign_add(
            self.execution_deferred.entry(reason).or_default(),
            units_consumed,
        )?;

        Ok(())
```

With `consume_execution`:

```
        if amount == 0 {
            return Ok(());
        }

        let units_consumed = checked_multiply(amount, multiplier)?;

        self.consume(units_consumed, self.execution_price())?;
        checked_assign_add(self.execution.entry(reason).or_default(), units_consumed)?;

        if self.cost_units_consumed >= self.check_point && !self.fully_repaid() {
            self.repay_all()?;
        }

        Ok(())
```

They're doing very different things - changing different entries, and having different logic around checking for repayment.

Reply:
To me, they both concern `execution` metering, with different time delays. Function bodies are implementation details.

Reply:
This is why we want `Decimal::round(mode)` rather than `round_to_zero`, `round_up` and `round_down`.

Reply:
As discussed, @dhedey's motivation is to disallow clients call `consume_deferred` during execution time, which otherwise may cause illegal state.

He is going to try to split the methods into two traits, to prevent kernel from accidentally calling it.